Installation in sphinx doc (#322)
lethe-cfd/lethe: doc/source/installation.rst

Installation
############

> Lethe can be installed on Linux, Mac OS and under the Windows Subsystem for Linux (WSL). Lethe has been tested on numerous distributions, including Ubuntu 18.04 LTS, Ubuntu 20.04 LTS, Centos 7 and Manjaro. Lethe requires a modern version of the deal.II library. At the time of this writing, ''deal.II 9.3'' and ''deal.II 10.0pre'' (the master branch version) are supported. The compatibility with these two branches is ensured by Continuous Integration (CI) using Github Actions. A `dealii fork <https://github.com/lethe-cfd/dealii>`_ is maintained by the Lethe organization. This fork does not include any modification to the library, but is the latest deal.II version against which Lethe was tested. We work hard to ensure compatibility with the latest deal.II version and we do not modify the library except through pull request on the official deal.II repository.

Reviewer: ```suggestion
Lethe can be installed on Linux, Mac OS and under the Windows Subsystem for Linux (WSL). Lethe has been tested on numerous distributions, including Ubuntu 18.04 LTS, Ubuntu 20.04 LTS, Centos 7 and Manjaro. Lethe requires a modern version of the deal.II library. At the time of this writing, ''deal.II 9.3'' and ''deal.II 10.0pre'' (the master branch version) are supported. The compatibility with these two branches is ensured by Continuous Integration (CI) using Github Actions. A `dealii fork <https://github.com/lethe-cfd/dealii>`_ is maintained by the Lethe organization. This fork does not include any modification to the library, but is the latest deal.II version against which Lethe was tested. We work hard to ensure compatibility with the latest deal.II version and we do not modify the library except through pull requests on the official deal.II repository.
```

The installation of Lethe consists in three major steps:

> 1. Installation of dealii dependency (mpi, numdiff, p4est, trilinos and METIS)

Reviewer: ```suggestion
1. Installation of deal.II dependencies (mpi, numdiff, p4est, trilinos and METIS)
```

> 2. Installation of the dealii library

Reviewer: ```suggestion
2. Installation of the deal.II library
```

3. Installation of lethe

> There are two methods to install the dependencies. They can be either installed manually or through ``candi``. The candi.sh shell script can download, configure, build, and install all the dependencies along with dealii. **This is the easiest way**, since it requires much less manual intervention. Even if you do not want to use candi to install deal.II, you can also use it the dependencies. This is by far the easiest way to proceed. On a single core computer with 8Gb of ram, count up to 8 hours for a first installation, and 3 hours for a dealii update. On a machine with 16 cores and 32 Gb or ram, this process will take less than an hour or so. The installation of deal.II and its dependencies (especially Trilinos), can be extremely ram consuming. Installation on a machine with less than 8Gb of ram is difficult at best, impossible at worst.

Reviewer: ```suggestion
There are two methods to install the dependencies. They can be either installed manually or through ``candi``. The candi.sh shell script can download, configure, build, and install all the dependencies along with deal.II. **This is the easiest way**, since it requires much less manual intervention. Even if you do not want to use candi to install deal.II, you can use it for the dependencies. This is by far the easiest way to proceed. On a single core computer with 8GB of RAM, count up to 8 hours for a first installation, and 3 hours for a deal.II update. On a machine with 16 cores and 32GB of RAM, this process will take less than an hour or so. The installation of deal.II and its dependencies (especially Trilinos), can be extremely RAM consuming. Installation on a machine with less than 8GB of RAM is difficult at best, impossible at worst.
```

> Lethe cannot be installed if deal.II has not been configured with p4est, Trilinos and METIS. Although Lethe can be run in serial and parallel (through mpi) mode, it uses p4est, METIS and Trilinos for mesh decomposition and the linear solvers respectively. 

Reviewer: ```suggestion
Lethe cannot be installed if deal.II has not been configured with p4est, Trilinos and METIS. Although Lethe can be run in serial and parallel mode (through mpi), it uses p4est, METIS and Trilinos for mesh decomposition and the linear solvers respectively. 
```

Installing deal.II using candi (Step #1)
-----------------------------------------

To install the dependencies (mpi, numdiff, p4est, trilinos and METIS) all together by Candi, the `procedure <https://github.com/dealii/candi.git>`_ on the candi repository can be followed.

Clone the candi git repository in a folder of your choice  (e.g. ``/home/username/software``). You can edit the ``candi.cfg`` file if you want to alter which dependencies are compiled. This can notably be used to force the installation of the deal.II master version instead of the current stable version by setting the ``STABLE_BUILD=false``

From the candi folder, the installation of candi can be launched using:

.. code-block:: text

  ./candi.sh -j $num_proc --prefix=$path

> where ``$num_proc`` is the number of threads you want to use to compile deal.II and ``$path$`` the installation prefix that is desired (e.g. ``/home/username/software/candi``). 

Author: ```suggestion
where ``$num_proc`` is the number of threads you want to use to compile deal.II and ``$path`` the installation prefix that is desired (e.g. ``/home/username/software/candi``). 
```

.. warning:: 
  For a computer with 8Gb of RAM, 1 thread (``num_proc=1``) should be used. For 16 Gb, 4 threads is reasonable. For 32 Gb, 16 threads or more can be used.

After installation, you should have a ``deal.ii-candi`` folder the installation prefix directory, with the dealii folder of the desired version (see section :ref:`update-dealii`), as well as the required dependencies (p4est, trilinos, etc.).

After installation, add an environment variable to your ``.bashrc`` either manually or through the following command:

.. code-block:: text

   echo "export DEAL_II_DIR=/home/username/deal.ii-candi/deal.II-<version>" >> ~/.bashrc`

> Installing Deal.II manually (Step #1)

Reviewer: ```suggestion
Installing deal.II manually (Step #1)
```

We first need to install the deal.II dependencies.

MPI
~~~~~

MPI, for Message Passing Interface, is required for the installation of all components of Lethe. Therefore it should be the first thing you install. On Debian Linux based OS, MPI can be installed directly from your package manager. 

As an example, in Ubuntu:

.. code-block:: text

  sudo apt-get install libopenmpi-dev openmpi-bin openmpi-common

In Manjaro or other arch based linux distribution:

.. code-block:: text
  
  sudo pacman -Sy openmpi

numdiff
~~~~~~~~

numdiff is used within the automatic testing procedure of Lethe to compare files obtained through floating point arithmetic. Without numdiff, Lethe automatic tests may fail when they should not. Numdiff can be installed directly from your package manager.

.. code-block:: text

  sudo apt-get install numdiff

Regrettably, numdiff is not available in the pacman package manager. It can be downloaded from the following `website <http://www.nongnu.org/numdiff/>`_. If you are using an arch distribution, we assume that you will already know how to carry on with the installation of numdiff. 

P4est
~~~~~~~

> To install p4est, the usual `installation procedure <https://www.dealii.org/current/external-libs/p4est.html>`_ of dealii can be followed.

Reviewer: ```suggestion
To install p4est, the usual `installation procedure <https://www.dealii.org/current/external-libs/p4est.html>`_ of deal.II can be followed.
```

Trilinos
~~~~~~~~~

> The installation of Trilinos should be done using the `installation procedure <https://www.dealii.org/current/external-libs/trilinos.html>`_ of dealii:

Reviewer: ```suggestion
The installation of Trilinos should be done using the `installation procedure <https://www.dealii.org/current/external-libs/trilinos.html>`_ of deal.II.
```

METIS
~~~~~~~

METIS is used for mesh partitioning for parallel computing purposes, specifically in cases with simplex grids. It can be downloaded `here <http://glaros.dtc.umn.edu/gkhome/metis/metis/download>`_ or through candi.

> Installation of dealii

Reviewer: ```suggestion
Installation of deal.II
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Clone deal.II from the `official repository <https://github.com/dealii/dealii>`_

.. code-block:: text

  git clone https://github.com/dealii/dealii 

Configure deal.II in a build folder at the same level as the source code

.. code-block:: text

  mkdir build
  cd build

Depending on how you have installed p4est, Trilinos and METIS, you may need to specify the installation folder of the three libraries

.. code-block:: text

>   cmake ../dealii -DDEAL_II_WITH_MPI=ON -DDEAL_II_WITH_TRILINOS=ON -DTRILINOS_DIR=path/to/your/trilinos/installation -DDEAL_II_WITH_P4EST=ON -DP4EST_DIR=path/to/your/p4est/installation  -DDEAL_II_WITH_METIS=ON -DMETIS_DIR=path/to/your/metis/installation -DCMAKE_INSTALL_PREFIX=/path/to/desired/installation`

Reviewer 2: Should we talk about ccmake? I always use this now and it feels easier

Author: I'm not sure i'd go there here for now. It's a good idea for a later addition, but then the installation become CMAKE instructions and etc. I'd rather keep it in the most simple format as possible.

Reviewer 2: That makes sense

> Compile dealii

Reviewer: ```suggestion
Compile deal.II
```

.. code-block:: text

  make -j<nprocessor> install

> Create an environment variable for the DEALII directory. 

Reviewer: ```suggestion
Create an environment variable for the deal.II directory
```

.. code-block:: text
 
  export DEAL_II_DIR=/path/to/dealii/installation

> It is generally recommended to add variable to your bashrc so it is always loaded:

Reviewer 2: ```suggestion
It is generally recommended to add the variable to your bashrc so it is always loaded:
```

.. code-block:: text

 echo "export DEAL_II_DIR=/path/to/dealii/installation" >> ~/.bashrc

.. _install-lethe:

Installation of lethe (Step #2)
-------------------------------

Clone lethe from the `official repository <https://github.com/lethe-cfd/lethe>`

> 

Reviewer 2: ```suggestion

.. code-block:: text

  git clone https://github.com/lethe-cfd/lethe 

```

Create a build folder at the same level as the lethe folder

.. code-block:: text

>   mkdir build

Reviewer 2: ```suggestion
  mkdir build
  cd build
```

> Compile Lethe choosing the compilation option (Debug or Release). You can also optionnaly specify and installation direction. We recommend that you do so, since this makes using Lethe much more comfortable.

Reviewer: ```suggestion
Compile Lethe choosing the compilation option (Debug or Release). You can also optionally specify a path to an installation directory of your choice. We recommend that you do so, since this makes using Lethe much more comfortable.
```

.. code-block:: text

  cmake ../lethe -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/home/username/path/to/installation

or

.. code-block:: text

  cmake ../lethe -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/home/username/path/to/installation

Then you can compile:

.. code-block:: text

  make -j<numprocs>

Testing your installation (Step #3)
-------------------------------------

Lethe comes pre-packaged with an extensive test suit for all of its modules. It can be used to test the validity of your installation. Within the build folder, the test suite can be launched with the following command:

.. code-block:: text

  ctest -j $numprocs

where $numprocs can be the number of physical cores on your machine.

.. _update-dealii:

Updating deal.II
-------------------

Through the git repository
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The deal.II version supported by Lethe is updated and tested every week or so, see the repository `here <https://github.com/lethe-cfd/dealii>`_. If Lethe was installed with this forked version of deal.II, updating your deal.II installation is as simple as pulling the repository and recompiling the deal.ii library. If your deal.II was installed manually using the deal.II master repository, the same process can be used.

Reviewer: ```suggestion
The deal.II version supported by Lethe is updated and tested every week or so, see the repository `here <https://github.com/lethe-cfd/dealii>`_. If Lethe was installed with this forked version of deal.II, updating your deal.II installation is as simple as pulling the repository and recompiling the deal.II library. If your deal.II was installed manually using the deal.II master repository, the same process can be used.
```
